[Ng Zi Xuan] iP

nzixuan opened the pull request:
### Duke Pro
>“Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

DukePro frees your mind of having to remember things you need to do. It's,

* text-based
* easy to learn
* ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2122s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
 - [ ] Reminders (coming soon)

---

If you Java programmer, you can use it to practice Java too. Here's the `main`
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

nzixuan committed:
Added basic Duke Skeleton

nzixuan committed:
Added add and list features

nzixuan committed:
Added task object and ability to mark as done

nzixuan committed:
Added Event, Todo and Deadline and respective functionalities

nzixuan committed:
Added Testing

nzixuan committed:
Refactored code to include exceptions and enums

nzixuan committed:
Added delete task

nzixuan committed:
Add save feature

nzixuan committed:
Add date time parsing to Deadline

nzixuan committed:
Merge branch 'branch-Level-7'

nzixuan committed:
Refactor code to include packages and more OOP

nzixuan committed:
Add JUnit Testing for Parser and Ui class

nzixuan committed:
Package application as JAR file

nzixuan committed:
Add JavaDoc for all public classes

nzixuan committed:
Refactored code to comply with java coding standards

nzixuan committed:
Add find functionality

nzixuan committed:
Merge branch 'branch-A-JavaDoc'

nzixuan committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/DukeException.java
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Storage.java
#	src/main/java/duke/TaskList.java
#	src/main/java/duke/Ui.java

nzixuan committed:
Add more JavaDoc to Command and Task packages

jovyntls commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Duke related Exceptions.
+ */
+public class DukeException extends Exception {
```
Maybe you could extend `RuntimeException` instead of `Exception` to be more specific :)

jovyntls commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,93 @@
+        }
+    }
+
+    void saveFile(ArrayList<Task> list) throws DukeException {
```
Is the method access modifier deliberately omitted here?

jiayi1129 replied:
Perhaps you can add a `public` access modifier to this method to make it clear that it is publicly accessible?

jovyntls commented on src/main/java/duke/task/Task.java:
```diff
@@ -0,0 +1,50 @@
+ * Can be marked as done and has a description of the task.
+ */
+public class Task {
+    private Boolean done;
```
```suggestion
    private Boolean isDone;
```
I saw that you use `isDone` in other places, is there a reason why `done` is used here instead?

KT27Learn replied:
Perhaps the done variable should be isDone? I've noticed that isDone is used in other places

jovyntls reviewed (commented):
I think the code is very well written and shows good use of OOP principles! I like the use of IntStream, StringBuilder and regex. 🥇

jiayi1129 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,90 @@
+ */
+public class Parser {
+
+    private enum Action {
+        Done,
+        Delete,
+        Todo,
+        Deadline,
+        Event,
+        Find,
+    }
```
I like how you separated the the possible user actions into enums! This makes it very clear what action the user is taking 👍

jiayi1129 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,90 @@
+package duke;
+
+import duke.command.*;
```
Perhaps you can separate out the package imports?

jiayi1129 commented on src/test/java/UiTest.java:
```diff
@@ -0,0 +1,39 @@
+                        "\tHello! I'm Duke.\n" +
+                        "\tWhat can I do for you?\n" +
+                        "\t____________________________________________________________\n"
+                ,out.toString());
```
Perhaps you can add a whitespace after the comma?

jiayi1129 commented on src/test/java/ParserTest.java:
```diff
@@ -0,0 +1,21 @@
+
+public class ParserTest {
+    @Test
+    public void parseBye() throws DukeException {
```
Perhaps you can consider naming your tests in the format `featureUnderTest_testScenario_expectedBehavior()`?

jiayi1129 reviewed (commented):
Overall, I like your program! I like the layers of abstraction starting from the main `Duke` driver class where the complexity of each components `Storage, Ui` etc are hidden from us. This makes your code very easy to read and understand. Also, your code is well commented and your descriptions are short, concise and yet still understood easily!

Other than a few minor nitpicks, LGTM! 😄

nzixuan committed:
Merge branch 'add-gradle-support' of https://github.com/nzixuan/ip into branch-A-Gradle

nzixuan committed:
Add gradle and checkstyle support

KT27Learn commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,63 @@
+
+    private final Storage storage;
+    private TaskList tasks;
+    private final Ui ui;
```
Final variable should be change to all caps

KT27Learn commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,95 @@
+ */
+public class Storage {
+
+    private final String filepath;
```
Perhaps the final variable count be all uppercase? I have noticed this in some other areas as well.

KT27Learn commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,95 @@
+        }
+    }
+
+    void saveFile(ArrayList<Task> list) throws DukeException {
```
I believe you missed out on the method access modifier to the method

KT27Learn reviewed (commented):
Everything looks great, and I think that it is an overall job well done! There are no major issues, and the code base was easy to understand. There are just some small styling pointers that was brought up.

nzixuan committed:
Refactor code for JavaFX UI.

nzixuan committed:
Add shadow jar plugin and support

Current jar files built do not have the libraries included when built.

Shadow plugin allows easy building of the project with the necessary libraries included.

Let's add support for the plugin and also fix issues with system fonts when building the jar files.

nzixuan committed:
Add assertions

Assumptions made while coding are not verified and can cause errors while running the code.

Assertions are simple to implement ways to ensure assumptions are met.

Let's add assertions in important junctures to make sure that code is ran without errors.

nzixuan committed:
Refactor code to be more inline with code quality standards

nzixuan committed:
Merge pull request #1 from nzixuan/branch-A-CodeQuality

Refactor code to be more inline with code quality standards

nzixuan committed:
Merge branch 'master' into branch-A-Assertions

nzixuan committed:
Merge pull request #2 from nzixuan/branch-A-Assertions

Add assertions

nzixuan committed:
Add Recurring Task Feature

nzixuan committed:
Tweak GUI to be more suitable for usage

nzixuan committed:
Update README.md

nzixuan committed:
Edit final changes to program logic

nzixuan committed:
Move README.md to documentation folder

nzixuan committed:
Move README.md to documentation folder

nzixuan committed:
Edit final changes to README.md and UI for release v1.0

nzixuan committed:
Edit documentation to correct errors in readme.md
